### PR TITLE
Add Simulated Player check (Standalone + PIE)

### PR DESCRIPTION
**Pull request by WVerlaek**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Adds a C++ and Blueprint accessible function to check if the player is a simulated player. It works both for PIE simulated player instances as well as when launched with a `-simulatedplayer` command line argument. This check should be used by developers to implement custom simulated player logic.

#### Release note
REQUIRED: Enter a simple message for this commit which will go in the release notes. Prefix with "Bugfix:" or "Feature:".
(Documentation changes are exempt)

#### Tests
How did you test these changes prior to submitting this pull request?
What automated tests are included in this PR?

STRONGLY SUGGESTED: How can this be verified by QA?

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

**Comment by WVerlaek** [comment]
I don't think the approach inside `IsSimulatedPlayer` to get the game instance is ideal, and don't know whether this works in all scenarios. `GWorld`'s documentation mentions that its usage should be avoided when possible, but other ways I tried for retrieving a reference to the GameInstance didn't work properly.

An alternative discussed with @Zomtorg offline is that we annotate the `IsSimulatedPlayer` function inside the `UGameInstance` engine class with `UFUNCTION(BlueprintPure)` (as seen [here](https://github.com/improbableio/UnrealEngine/blob/b36c8ceca978a5c2f96a5ad35977fb982b362d54/Engine/Source/Runtime/Engine/Classes/Engine/GameInstance.h#L388)) such that from blueprints you can call `GetGameInstance()->IsSimulatedPlayer()` directly, and that we don't add an `IsSimulatedPlayer` method inside the GDK. You would need access to the GameInstance in order to call `IsSimulatedPlayer`, but this is available from blueprints and actor classes. 

Any thoughts on this?

File: SpatialGDK/Source/SpatialGDK/Public/SimulatedPlayers/SimulatedPlayer.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#include "Kismet/BlueprintFunctionLibrary.h"
+#include "SimulatedPlayer.generated.h"
+
+UCLASS()
+class SPATIALGDK_API USimulatedPlayer : public UBlueprintFunctionLibrary
+{
+	GENERATED_BODY()
+
+public:
+	/** Get whether this client is simulated.
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/SimulatedPlayers/SimulatedPlayer.h]: Nit - drop this one line down. The all text should be on a line starting with a single *

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+#include "Misc/Parse.h"
+#include "Misc/CommandLine.h"
+#include "SpatialConstants.h"
+
+DEFINE_LOG_CATEGORY_STATIC(LogSimulatedPlayer, Log, All);
+
+bool USimulatedPlayer::IsSimulatedPlayer()
+{
+#if WITH_EDITOR
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: Why do we use the preprocessor for this rather than a check?

**Comment by WVerlaek** [comment]
Refactored the `IsSimulatedPlayer` implementation to use a WorldContextObject instead to get a reference to the game instance. For blueprints this doesn't change anything: `IsSimulatedPlayer` can still be called without passing any arguments, the WorldContextObject is implicitly passed from the blueprint. For C++ usage you can pass a reference to (almost) any `UObject` that is present in your world as a WorldContextObject. E.g. when calling within an Actor class you can call `IsSimulatedPlayer(this)`

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp (hunk 2 above)
Comment (WVerlaek) [reply SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: That's gone now 👍

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	}
+
+	UGameInstance* gameInstance = world->GetGameInstance();
+	if (gameInstance == nullptr)
+	{
+		UE_LOG(LogSimulatedPlayer, Error, TEXT("Cannot check if player is simulated: game instance is null. Defaulting to false."));
+		return false;
+	}
+
+	return gameInstance->IsSimulatedPlayer();
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: Is this function not part of this PR?

File: SpatialGDK/Source/SpatialGDK/Public/SimulatedPlayers/SimulatedPlayer.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "Kismet/BlueprintFunctionLibrary.h"
+#include "SimulatedPlayer.generated.h"
+
+UCLASS()
+class SPATIALGDK_API USimulatedPlayer : public UBlueprintFunctionLibrary
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Public/SimulatedPlayers/SimulatedPlayer.h]: Make this class name more reflective of what it does. e.g. `SimPlayerBPFunctionLibrary`

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SimulatedPlayer.h"
+
+#include "Engine.h"
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: See if you can get away with not including the whole `Engine.h`. `EngineMinimal.h` maybe.

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+#include "Misc/Parse.h"
+#include "Misc/CommandLine.h"
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: These includes shouldn't be necessary

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+#include "SpatialConstants.h"
+
+DEFINE_LOG_CATEGORY_STATIC(LogSimulatedPlayer, Log, All);
+
+bool USimulatedPlayer::IsSimulatedPlayer(UObject* WorldContextObject)
+{
+	UEngine* engine = GEngine;
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: `Engine` not `engine` according to UE4 naming conventions.

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	if (engine == nullptr)
+	{
+		UE_LOG(LogSimulatedPlayer, Error, TEXT("Cannot check if player is simulated: GEngine is null. Defaulting to false."));
+		return false;
+	}
+
+	UWorld* world = engine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::ReturnNull);
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: Error handling in this function is too heavy handed. Refactor please.

`GEngine` better be non-null otherwise we have bigger problems. So just do what `UGameplayStatics::GetGameInstance()` does, and in failure print a `Warning` not `Error`.

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	if (world == nullptr)
+	{
+		UE_LOG(LogSimulatedPlayer, Error, TEXT("Cannot check if player is simulated: world returned by WorldContextObject is null. Defaulting to false."));
+		return false;
+	}
+
+	UGameInstance* gameInstance = world->GetGameInstance();
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: `GameInstance`

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp (hunk 5 above)
Comment (WVerlaek) [reply SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: Nevermind, include is no longer necessary after refactor

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp (hunk 6 above)
Comment (WVerlaek) [reply SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimulatedPlayer.cpp]: Thanks, removed

File: SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimPlayerBPFunctionLibrary.cpp [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "Engine/GameInstance.h"
+#include "Kismet/GameplayStatics.h"
+
+DEFINE_LOG_CATEGORY_STATIC(LogSimulatedPlayer, Log, All);
+
+bool USimPlayerBPFunctionLibrary::IsSimulatedPlayer(UObject* WorldContextObject)
+{
+	UGameInstance* GameInstance = UGameplayStatics::GetGameInstance(WorldContextObject);
+	if (GameInstance == nullptr) {
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/SimulatedPlayers/SimPlayerBPFunctionLibrary.cpp]:
```suggestion
	if (GameInstance == nullptr) 
	{
```

File: SpatialGDK/Source/SpatialGDK/Public/SimulatedPlayers/SimPlayerBPFunctionLibrary.h [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+	GENERATED_BODY()
+
+public:
+	/**
+	 * Get whether this client is simulated.
+	 * This will return true for clients launched inside simulated player deployments,
+	 * or simulated clients launched from the Editor.
+	 */
+	UFUNCTION(BlueprintPure, Category="SpatialOS|SimulatedPlayer", meta = (WorldContext = WorldContextObject))
+	static bool IsSimulatedPlayer(UObject* WorldContextObject);
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Public/SimulatedPlayers/SimPlayerBPFunctionLibrary.h]:
```suggestion
	static bool IsSimulatedPlayer(const UObject* WorldContextObject);
```

**Comment by WVerlaek** [comment]
Closed this PR, all Simulated Player work is included in https://github.com/spatialos/UnrealGDK/pull/932
